nixos/modules/services/databases/openldap.nix

```diff
@@ -244,57 +356,156 @@ in
-
   config = mkIf cfg.enable {
-    assertions = [
-      {
-        assertion = cfg.configDir != null || cfg.rootpwFile != null || cfg.rootpw != null;
-        message = "services.openldap: Unless configDir is set, either rootpw or rootpwFile must be set";
-      }
-    ];
+    warnings = let
+      deprecations = [
+        { old = "logLevel"; new = "attrs.olcLogLevel"; }
+        { old = "defaultSchemas";
+          new = "children.\"cn=schema\".includes";
+          newValue = "[\n    ${lib.concatStringsSep "\n    " [
+            "\${pkgs.openldap}/etc/schema/core.ldif"
+            "\${pkgs.openldap}/etc/schema/cosine.ldif"
+            "\${pkgs.openldap}/etc/schema/inetorgperson.ldif"
+            "\${pkgs.openldap}/etc/schema/nis.ldif"
+          ]}\n  ]"; }
+        { old = "database"; new = "children.\"cn={1}${cfg.database}\""; newValue = "{ }"; }
+        { old = "suffix"; new = "children.\"cn={1}${cfg.database}\".attrs.olcSuffix"; }
+        { old = "dataDir"; new = "children.\"cn={1}${cfg.database}\".attrs.olcDbDirectory"; }
+        { old = "rootdn"; new = "children.\"cn={1}${cfg.database}\".attrs.olcRootDN"; }
+        { old = "rootpw"; new = "children.\"cn={1}${cfg.database}\".attrs.olcRootPW"; }
+        { old = "rootpwFile";
+          new = "children.\"cn={1}${cfg.database}\".attrs.olcRootPW";
+          newValue = "{ path = \"${cfg.rootpwFile}\"; }";
+          note = "The file should contain only the password (without \"rootpw \" as before)"; }
+      ];
+    in (optional (cfg.extraConfig != "" || cfg.extraDatabaseConfig != "") ''
+      The options `extraConfig` and `extraDatabaseConfig` of `services.openldap`
+      are deprecated. This is due to the deprecation of `slapd.conf`
+      upstream. Please migrate to `services.openldap.settings`.
+
+      After deploying this configuration, you can run:
+        slapcat -F ${configDir} -n0 -H 'ldap:///???(!(objectClass=olcSchemaConfig))'
+      on the same host to print your current configuration in LDIF format,
+      which should be straightforward to convert into Nix settings.
+    '') ++ (flatten (map (args@{old, new, ...}: lib.optional ((lib.hasAttr old cfg) && (lib.getAttr old cfg) != null) ''
+      The attribute `services.openldap.${old}` is deprecated. Please set it to
+      `null` and use the following option instead:
+
+        services.openldap.settings.${new} = ${args.newValue or (
+          let oldValue = (getAttr old cfg);
+          in if (isList oldValue) then "[ ${concatStringsSep " " oldValue} ]" else oldValue
+        )}
+    '') deprecations)) ++ (optional (cfg.configDir != null && (versionOlder config.system.stateVersion "20.09")) ''
+      The attribute `services.openldap.settings` now exists, and may be more
+      useful than `services.openldap.configDir`. If you continue to use
+      `configDir`, ensure that `olcPidFile` is set to "/run/slapd/slapd.pid".
+
+      Set `system.stateVersion` to "20.09" or greater to silence this message.
```

Comment:
`stateVersion` can't just be changed without potentially breaking other things in the process, so I don't think it's a good idea to recommend this. I think the next best way to inform users of new features is via the release notes.

Reply:
OK, added some release notes. This will be presumably be in for `21.03`.